File: Leet 297/solution.cs
using System.Text;

public class TreeNode(int x)
{
    public int val = x;
    public TreeNode? left = null;
    public TreeNode? right = null;
}

public class Codec
{
    // Encodes a tree to a single string.
    public static string Serialize(TreeNode root)
    {
        if (root == null)
            return "";

        Queue<TreeNode> que = new();
        que.Enqueue(root);

        StringBuilder sb = new();
        sb.Append(root.val);

        while (que.Count > 0)
        {
            int count = que.Count;

            for (int i = 0; i < count; i++)
            {
                var current = que.Dequeue();

                sb.Append(',');
                if (current.left != null)
                {
                    que.Enqueue(current.left);
                    sb.Append(current.left.val);
                }
                else
                    sb.Append("null");

                sb.Append(',');
                if (current.right != null)
                {
                    que.Enqueue(current.right);
                    sb.Append(current.right.val);
                }
                else
                    sb.Append("null");
            }
        }

        // Remove trailing nulls and commas
        string result = sb.ToString();
        while (result.EndsWith(",null"))
            result = result.Substring(0, result.Length - 5);

        return result;
    }

    private static bool TryGetValue(string[] values, int idx, out int result)
    {
        result = 0;
        if(idx >= values.Length || values[idx] == "null")
            return false;
        return int.TryParse(values[idx], out result);
    }

    // Decodes your encoded data to tree.
    public static TreeNode Deserialize(string data)
    {
        if (string.IsNullOrEmpty(data))
            return null;

        string[] values = data.Split(',');
        int idx = 0;

        if (!TryGetValue(values, idx++, out int rootVal))
            return null;

        TreeNode root = new(rootVal);
        Queue<TreeNode> que = new();
        que.Enqueue(root);

        while (que.Count > 0 && idx < values.Length)
        {
            var current = que.Dequeue();


            if (TryGetValue(values, idx++, out int leftVal))
            {
                current.left = new(leftVal);
                que.Enqueue(current.left);
            }

            if (TryGetValue(values, idx++, out int rightVal))
            {
                current.right = new(rightVal);
                que.Enqueue(current.right);
            }
        }

        return root;
    }
}


class Program
{
    // helper to create a simple tree
    private static TreeNode Root()
    {
        TreeNode root = new(1);
        root.left = new(2);
        root.right = new(3);
        root.right.left = new(4);
        root.right.right = new(5);
        return root;
    }

    private static void PrintTree(TreeNode root)
    {
        if(root == null) return;

        Stack<TreeNode> stk = new();
        stk.Push(root);

        while (stk.Count > 0)
        {
            var current = stk.Pop();
            
            if (current == null)
                continue;

            Console.WriteLine(current.val);

            stk.Push(current.left);
            stk.Push(current.right);
        }
    }

    private static void Main()
    {
        Console.WriteLine("297. Serialize and Deserialize Binary Tree");
        
        var serialize = Codec.Serialize(Root());
        Console.WriteLine(serialize);

        var root = Codec.Deserialize(serialize);
        PrintTree(root);
    }
}

File: Leet 297/solution.go
package main

import (
	"fmt"
	"strconv"
	"strings"
)

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// Queue using two stacks
type Queue struct {
	stackA []*TreeNode
	stackB []*TreeNode
}

func (q *Queue) Enqueue(node *TreeNode) {
	q.stackA = append(q.stackA, node)
}

func (q *Queue) Dequeue() *TreeNode {
	// move values from a to b
	if len(q.stackB) == 0 {
		for len(q.stackA) > 0 {
			var current *TreeNode = q.stackA[len(q.stackA)-1]
			q.stackA = q.stackA[:len(q.stackA)-1]

			q.stackB = append(q.stackB, current)
		}
	}

	// pop result
	item := q.stackB[len(q.stackB)-1]
	q.stackB = q.stackB[:len(q.stackB)-1]

	return item
}

func (q *Queue) IsEmpty() bool {
	return len(q.stackA) == 0 && len(q.stackB) == 0
}

func (q *Queue) Count() int {
	return len(q.stackA) + len(q.stackB)
}

func CodecSerialize(root *TreeNode) string {
	if root == nil {
		return ""
	}

	que := Queue{}
	que.Enqueue(root)

	var sb strings.Builder
	sb.WriteString(strconv.Itoa(root.Val))

	for !que.IsEmpty() {
		count := que.Count()

		for count > 0 {
			count--

			current := que.Dequeue()

			sb.WriteString(",")
			if current.Left != nil {
				que.Enqueue(current.Left)
				sb.WriteString(strconv.Itoa(current.Left.Val))
			} else {
				sb.WriteString("null")
			}

			sb.WriteString(",")
			if current.Right != nil {
				que.Enqueue(current.Right)
				sb.WriteString(strconv.Itoa(current.Right.Val))
			} else {
				sb.WriteString("null")
			}
		}
	}

	var result string = sb.String()

	// trim repeating ",null" from the end of the result string
	var target string = ",null"
	var targetLen int = len(target)
	var resultEnd int = len(result)
	for resultEnd >= targetLen && result[resultEnd-targetLen:resultEnd] == target {
		resultEnd -= targetLen
	}

	return result[:resultEnd]
}

// helper function that trys to convert a value from values at idx to an int
//
// the []string values consists of numbers or null
func tryGetValue(values []string, idx int) (int, bool) {
	if idx < 0 || idx >= len(values) || values[idx] == "null" {
		return -1, false
	}

	val, err := strconv.Atoi(values[idx])
	if err != nil {
		return -1, false
	}

	return val, true
}

func CodecDeserialize(data string) *TreeNode {
	if data == "" {
		return nil
	}

	var values []string = strings.Split(data, ",")
	idx := 0

	rootVal, ok := tryGetValue(values, idx)
	if !ok {
		return nil
	}

	root := new(TreeNode)
	root.Val = rootVal

	que := Queue{}
	que.Enqueue(root)
	idx += 1

	for que.Count() > 0 && idx < len(values) {
		current := que.Dequeue()

		leftVal, ok := tryGetValue(values, idx)
		if ok {
			current.Left = new(TreeNode)
			current.Left.Val = leftVal
			que.Enqueue(current.Left)
		}
		idx++

		rightVal, ok := tryGetValue(values, idx)
		if ok {
			current.Right = new(TreeNode)
			current.Right.Val = rightVal
			que.Enqueue(current.Right)
		}
		idx++
	}

	return root
}

func printTree(root *TreeNode) {
	if root == nil {
		return
	}

	fmt.Println(root.Val)
	printTree(root.Left)
	printTree(root.Right)
}

func createTree() *TreeNode {
	root := new(TreeNode)
	root.Val = 1

	root.Left = new(TreeNode)
	root.Left.Val = 2

	root.Right = new(TreeNode)
	root.Right.Val = 3

	root.Right.Left = new(TreeNode)
	root.Right.Left.Val = 4

	root.Right.Right = new(TreeNode)
	root.Right.Right.Val = 5
	return root
}

func main() {
	fmt.Println("297. Serialize and Deserialize Binary Tree")

	var root *TreeNode = createTree()

	var sRoot string = CodecSerialize(root)
	fmt.Println(sRoot)

	var dRoot *TreeNode = CodecDeserialize(sRoot)
	printTree(dRoot)
}
